docs/proposals/resource-quota.md

```diff
@@ -0,0 +1,404 @@
+creation rate.  For example, a replication controller that keeps trying to
+scale up a resource may deplete the user's daily creation rate.  Controller's
+should be sensitive to quota by not making requests that would just be rejected
+by quota.  For example, if a quota allowed 10 pods in total, but the user set
+a replication controller replicaSet to 100, and the admin had a daily creation
+rate of 1000, the controller would rapidly deplete the creation rate.  If we
+target a creation rate to a particular service account or user-identity that is
+separate from the controllers, better results could potentially be achieved.
+
+Phase 4:
+
+* Support ability for namespace-admin to quota based on local understanding of
+workload
+ * Add `LabelSelector` to `v1` API
+ * Add `LabelSelector` to `ResourceQuotaSpec` to enable scoping on matched labels
+
+One of the interesting challenges with support a `LabelSelector` based
+tracking system is that the set of labels are fluid.  Unlike the scenarios
+presented around `FieldSelector`, the set of labels can change at any
+point over the life of the object.  As a consequence of this, one of the
+things that you must be careful to avoid is double-counting the same
+resource for a particular quota.  A potential strategy to avoid double
+counting resources that have been labeled differently post creation
+is to keep a ledger of resources that a `ResourceQuota` has charged
+to its quota document.  The ledger could be stored as an annotation
+of `APIVersion`, `Kind`, `Name` records.  If a particular resource
+had been tracked by quota, it would have an entry in the ledger.  A quota
+document would be able to track and account for double-counting resources
+by checking if the resource was in its known ledger.  If so, it would be
+able to properly distinguish between charging for new resources or updating
+resources that have previously been charged to the quota as a resources labels
+change.  The assumption in this solution is that the ability to edit a quota
+is very limited by other resources in the system.  If annotations prove
+insufficient, a `ResourceQuotaStatus` could maintain the ledger.  A general
+ledger is an interesting topic to explore in future scenarios.
+
+Phase 5:
+
+* investigate feeding metrics into quota controller that let it know what
+namespaces have had resources used to better target how quota controller
+can replenish quota.
+ * what is done here is a factor of how much we can scale the quota
+ controller by watching for more resource types, and how long we can
+ wait before doing a full recalculation, TBD based on phase 1
+
+## Example scenario
+
+### Administrator running large cluster that needs to control footprint
+
+As a cluster admin, I want to restrict each namespace to following:
+
+* no more than 4 pods, 10 replication controllers, 10 jobs, 10 services
+* no more than 2 "job" pods active at any time
+* no more than 2 "long-running" pods active at any time
```

Comment:
The discussion isn't completely tangential. Our plan for multiple schedulers in a nutshell has been (note that this is described moving from back to front of the pipeline)
1) Have schedulers use the value of the annotation with key `scheduler.alpha.kubernetes.io/name` to decide whether they are responsible for a pod. This is already implemented.
2) Implement a generic admission controller, driven by a policy object called MetadataPolicy (previously PodPolicy) that specifies what labels/annotations to write on an object based on various fields that already exist on the object. Our hope (as you can tell from the new name) was that it could just look at metadata (labels and annotations in particular) and thus could be generic across objects. BTW it would also implement required, prohibited, and defaulting for labels/annotations. In the case of multiple schedulers scheduler, it would look at a label or annotation that indicates QoS, and apply its policy to create a `scheduler.alpha.kubernetes.io/name` annotation with appropriate value. This is not implemented yet.
3) Have a separate admission controller that writes the QoS label/annotation based on properties of the pod. This is not implemented yet.

You can replace QoS with long-running vs. not, and the idea would be the same.

I think this can still work in the scheme you are describing but it's unfortunate we won't be able to reuse the machinery you will build to, for example, write QoS into PodStatus. If it wrote an annotation instead of PodStatus then it would serve as (3) and we wouldn't need to build (3). Analogously if long-running vs. not is not exposed as an annotation -- our approach will still work but we will still need to build it into (3).